napari/_qt/containers/qt_layer_tree.py

```diff
@@ -0,0 +1,130 @@
+        ):  # the "checked" state of this item
+            if layer.visible:
+                parents_visible = all(p._visible for p in layer.iter_parents())
+                return Qt.Checked if parents_visible else Qt.PartiallyChecked
+            else:
+                return Qt.Unchecked
+        if role == Qt.ItemDataRole.SizeHintRole:  # determines size of item
+            return QSize(200, 34)
+        if role == ThumbnailRole:  # return the thumbnail
+            thumbnail = layer.thumbnail
+            return QImage(
+                thumbnail,
+                thumbnail.shape[1],
+                thumbnail.shape[0],
+                QImage.Format_RGBA8888,
+            )
+        # normally you'd put the icon in DecorationRole, but we do that in the
+        # # LayerDelegate which is aware of the theme.
+        # if role == Qt.ItemDataRole.DecorationRole:  # icon to show
+        #     pass
+        return super().data(index, role)
+
+    def setData(
+        self,
+        index: QModelIndex,
+        value: Any,
+        role: int = Qt.ItemDataRole.EditRole,
+    ) -> bool:
+        if role == Qt.ItemDataRole.CheckStateRole:
+            self.getItem(index).visible = value
+        elif role == Qt.ItemDataRole.EditRole:
+            self.getItem(index).name = value
+            role = Qt.ItemDataRole.DisplayRole
+        else:
+            return super().setData(index, value, role=role)
+
+        self.dataChanged.emit(index, index, [role])
+        return True
+
+    def _process_event(self, event):
+        # The model needs to emit `dataChanged` whenever data has changed
+        # for a given index, so that views can update themselves.
+        # Here we convert native events to the dataChanged signal.
+        if not hasattr(event, 'index'):
+            self.dataChanged.emit(
+                self.index(0), self.index(self.rowCount()), []
+            )
+            return
+        role = {
+            'thumbnail': ThumbnailRole,
+            'visible': Qt.ItemDataRole.CheckStateRole,
+            'name': Qt.ItemDataRole.DisplayRole,
+        }.get(event.type)
+        roles = [role] if role is not None else []
+        top = self.nestedIndex(event.index)
+        bot = self.index(top.row() + 1)
+        self.dataChanged.emit(top, bot, roles)
+
+
+class QtLayerTreeView(QtNodeTreeView):
+    _root: LayerGroup
+    model_class = QtLayerTreeModel
+
+    def __init__(self, root: LayerGroup, parent: QWidget = None):
+        super().__init__(root, parent)
+        self.setItemDelegate(LayerDelegate())
+        self.setIndentation(18)
+
+        # couldn't set in qss for some reason
+        font = self.font()
+        font.setPointSize(12)
+        self.setFont(font)
+
+        # # This reverses the order of the items in the view,
+        # # so items at the end of the list are at the top.
+        # # FIXME! not working at the moment, and causing test segfaults
+        # self.setModel(ReverseProxyModel(self.model()))
```

Comment:
Following https://github.com/napari/napari/issues/6345#issuecomment-1900165860 gave that a check and I think the segfault is caused due to the logic inside the proxy model `dropMimeData` since, as it is, it doesn't handle parent change and parent mapping logic that is needed when using a tree view/model. So, when using a tree view, seems like there is a case where both row and column args are received with a -1 value. That indicates a parent/root element change. I think defining and using a proxy model class like the following could help here:

```python
class ReverseProxyTreeModel(QSortFilterProxyModel):
    """Proxy Model that reverses the view order of a _BaseEventedItemModel with tree like structure features."""

    def __init__(self, model: _BaseEventedItemModel) -> None:
        super().__init__()
        self.setSourceModel(model)
        self.setSortRole(SortRole)
        self.sort(0, Qt.SortOrder.DescendingOrder)

    def dropMimeData(self, data, action, destRow, col, parent):
        """Handle destination row for dropping with reversed indices and parent index change."""
        source_parent = self.mapToSource(parent)
        if source_parent.isValid():
            if destRow == -1 and col == -1:
                return self.sourceModel().dropMimeData(data, action, destRow, col, source_parent)
            row = 0 if destRow == -1 else self.sourceModel().rowCount() - destRow
            return self.sourceModel().dropMimeData(data, action, row, col, source_parent)
        return False
```

![layergroup_drapdrop](https://github.com/napari/napari/assets/16781833/52c96296-1248-425a-bcc9-22620073af83)

Besides the logic above, probably some other validations could be needed in the `dropMimeData` method depending on the drag and drop behavior that needs to be supported.

Some info in the Qt docs about the tree view behavior with the `dropMimeData` args is at: https://doc.qt.io/qt-6/qabstractitemmodel.html#dropMimeData